### PR TITLE
[public-api] Auto-generate proxy handler

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

**What**
This PR introduces codegen for a proxy handler. The handler is responsible for accepting a request (as JSON, or gRPC) and forwards that payload onto an underlying grpc client. This effectively results in proxying the requests.

**Why**
We want to be able to accept all incoming requests over JSON/gRPC on Public API, but delegate to grpc endpoints on `server` without having to manually wrap and un-wrap the objects.
This allows us to have a hybrid implementation between server and public-api while we're in the transitionary period.

**How to use it**
When we want to proxy a service, we use the generated proxy handler when registering the route. In code, this looks like:

For example, to proxy a service (using ide-service as an example)
```golang
// see public-api-server/pkg/server/server.go #register method
	projectsRoute, projectsServiceHandler := v1connect.NewProjectsServiceHandler(apiv1.NewProjectsService(deps.connPool), handlerOptions...)
	rootHandler.Mount(projectsRoute, projectsServiceHandler)

	oidcRoute, oidcServiceHandler := v1connect.NewOIDCServiceHandler(apiv1.NewOIDCService(deps.connPool, deps.expClient, deps.dbConn, deps.cipher), handlerOptions...)
	rootHandler.Mount(oidcRoute, oidcServiceHandler)

	fooRoute, fooHandler := v1connect.NewIDEClientServiceHandler(&v1connect.ProxyOIDCServiceHandler{Client: grpcClient })
```

## Related Issue(s)

Fixes #

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-slow-database
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		g.P("type " + handlerStructName + " struct {")
+		g.P("	Client " + g.QualifiedGoIdent(file.GoImportPath.Ident(service.GoName+"Client")))
+		g.P("	Unimplemented" + service.GoName + "Handler")
+		g.P("}")
+		g.P()
+
+		for _, method := range service.Methods {
+			// We do not generate any non-unary methods, for now.
+			// Should we need these, we can choose to do so and handle them explicitly.
+			// The handler still continues to work fine, as it inhertis from the default Unimplemented handling, and will
```
Comment (csweichel) [inline components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go:79]: typo
```suggestion
			// The handler still continues to work fine, as it inherits from the default Unimplemented handling, and will
```

File: components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	g.Import(file.GoImportPath)
+	g.P()
+
+	// generate individual services
+	for _, service := range file.Services {
+		// generate struct definition
+		handlerStructName := "Proxy" + service.GoName + "Handler"
+
+		// Generate a type assertion to ensure the handler implements the connect handler interface
+		g.P("var _ " + service.GoName + "Handler" + " = (*" + handlerStructName + ")(nil)")
```
Comment (csweichel) [inline components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go:67]: nit: for consistencies sake fmt.Sprintf would be preferable throughout

File: components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+				method.GoName,
+				g.QualifiedGoIdent(contextPackage.Ident("Context")),
+				g.QualifiedGoIdent(connectPackage.Ident("Request"))+"["+g.QualifiedGoIdent(method.Input.GoIdent)+"]",
+				g.QualifiedGoIdent(connectPackage.Ident("Response"))+"["+g.QualifiedGoIdent(method.Output.GoIdent)+"]",
+			))
+
+			// method implementation
+			g.P(fmt.Sprintf("	resp, err := s.Client.%s(ctx, req.Msg)", method.GoName))
+			g.P("	if err != nil {")
+			g.P("		// TODO(milan): Convert to correct status code")
```
Comment (csweichel) [inline components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go:97]: Q: What conversion needs to happen here? Isn't the status code the same for gRPC and connect-go?

Comment (easyCZ) [reply components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go:97]: I need to test this, but from the looks we need to covert the `gRPC` specifics of the error back to `connect` Errors, such that they serialize correctly over to the HTTP response code and headers.

File: components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go (hunk 2 above)
Comment (easyCZ) [reply components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go]: done

File: components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+			if !f.Generate {
+				continue
+			}
+			generateFile(gen, f)
+		}
+		return nil
+	})
+}
+
+func generateFile(gen *protogen.Plugin, file *protogen.File) {
```
Comment (geropl) [inline components/public-api/go/protoc-proxy-gen/protoc-proxy-gen.go:34]: Really love how compact this is :heart_eyes:

**Review by geropl — APPROVED** [review]
Really awesome! Let's give this a try, and improve as we go.

**Pull request merged**